Allow rule based operations on existing collections.

jmchilton opened the pull request:
Allow the rules DSL & GUI component (#5365) to operate on existing collections - this enables very flexible filtering, sorting, relabelling, grouping, flattening, and general re-organization of existing collections (e.g. the outputs of tools). This is implemented as a collection operation tool so that it should be executable interactively in the tool form and in a batch fashion as part of workflow executions.

For this to be tracked properly as a tool execution and to work properly in the tool form, I've implemented a new tool framework parameter type (including tool form integration) called "rules". Right now there is just one tool that uses this parameter type - but one can imagine other operations using the rule builder - e.g. creating collections from samplesheets as the first step in an automated workflow using the tool framework instead of the upload API.

This can thought of as a more GUI friendly alternative to some of my proposed collection operations (#2496 , #1313) in the past that consumed JavaScript expressions. This alternative approach is much easier to visualize for GUI based manipulation and is simpler than learning JavaScript if one does not know how to program, but the ubiquity of JavaScript would provide its owns benefits. Hopefully in the long run both approaches prove useful within the Galaxy ecosystem.

- [X] ~Outline the collection operation tool and parameter plumbing. Backend basics. (1 Day)~
- [X] ~Implement simplest tool form parameter. (1 Day)~
- [X] ~Connect tool form parameter to rule based builder component, including callback. (1 Day)~
- [X] ~Simplest rule working in API to create flat and nested collections. (1 Days)~
- [X] ~Simplest rule working GUI.~ (1/2 Day)
- [X] ~Preload collection identifiers as columns the first time the rules widget is loaded.~
- [x] ~Selenium test case for no-op tool. (1 Day)~
- [X] ~Flatten use case working API (1/2 Day)~
- [X] ~Flatten use case - GUI + Selenium. (1/2 Day)~
- [x] ~Swap the frontend to use Python-like regexes so these two approaches give the same results. (2 Days)~
- [x] ~Develop specification framework for rules DSL that can be used to test frontend JS and backend Python.~
- [x] Backend implementation of remaining rule types: (2 Days)
  - [x] add_column_basename
  - [x] add_column_rownum
  - [x] add_column_value
  - [x] add_column_metadata
  - [x] add_column_regex
  - [x] add_column_concatenate
  - [x] add_column_substr
  - [x] remove_columns
  - [x] add_filter_regex
  - [x] add_filter_count
  - [x] add_filter_empty
  - [x] add_filter_matches
  - [x] add_filter_compare
  - [x] sort
  - [x] swap_columns
  - [x] split_columns
- [x] ~Rework output collection type of this tool - it can be anything but it is determinable based on the content of the rules parameter - the workflow editor needs to be able to leverage this information. (2 Days)~ 
   - [x] ~Changing the collection type in the editor breaks the workflow connection - this is better than nothing but it'd be good if there was a warning or if it would attempt to re-establish the connection and alter the mapping if needed.~ (Added a async warning about this in the editor - obviously it'd be better if we had a more permanent warning mechanism for workflow changes but the UI framework isn't there yet.)
- [X] ~Testing rules in workflows in the API. 1 Day.~
- [X] ~Testing rules editor in workflow editor in the GUI (selenium).~ 1 Day.
- [x] ~Conformance tests for the rules language and a way to run them against both the Python and ES6 implementations (attempt to get regex working same in JS and Python). This is a tricky aspect of doing this on the backend this way.~ 2 Days
- [x] ~Testing run rules tool in workflows in the GUI (selenium).~ 1 Day.
- [X] ~Prevent workflow runtime edit of this parameter.~
- [x] Pretty validation on the backend and frontend. 1 Day.
- [x] Pretty display of rules in the...
    - [X] Tool form.
    - [x] Workflow summary.

xref https://github.com/galaxyproject/galaxy/issues/5381 (third bullet point)

jmchilton committed:
Fix for split columns rule definition.

jmchilton committed:
Add simple help for apply rules tool.

mvdbeek committed:
Add apply rules tool to standard tool xml sample

jmchilton committed:
RuleBuilder tool param fix for to_dict.

An exception in generating a "text representation" of a None value was preventing the rest of to_dict from running properly for that parameter value. This restores the rule preview table in the tool form.

jmchilton committed:
Fix modal title for apply_rules branch changes.

mvdbeek commented on lib/galaxy/tools/apply_rules.xml:
```diff
@@ -0,0 +1,36 @@
+may be used in workflows as well where no such preview can be generated.
+
+This tool is an advanced feature but has a lot of flexibility - it can be used to process collections
+with arbitrary nesting and can do many be used to do many kinds of filtering, re-sorting, nesting, 
```
s/many be used to do// ?

jmchilton committed:
Fix up help language in apply_rules tool.

mvdbeek commented:
This is very nice, I really like the flexibility this offers for reshaping collections. Certainly the flexible filtering will be very helpful for nested collections and workflows where there is a common part and a sample / condition specific part. This all works very nicely!

martenson reviewed (approved):
This PR presents a powerful functionality that takes dataset hierarchy and history manipulation to the next level and I am eager to see what impressive things the community can do with it.

The thing I am a bit concerned about is the user experience since the interface is unlike any we use currently or that is widely adopted. Now we have it in three places in Galaxy.

I propose we focus on gathering feedback on this throughout the 18.05 release with the aim to improve user intuition of building and managing these rules. 

Together with https://github.com/galaxyproject/training-material/pull/676 this is already very useful and ready to be merged.

jmchilton committed:
Allow rule based operations on existing collections.

Allow the rules DSL & GUI component to operate on existing collections to allow filtering, sorting, modifying identifiers and general re-organization of existing collections (e.g. the outputs of tools). Implementing this as a collection operation tool so that it should be executable interactively in the tool form and in a batch fashion as part of workflow executions.

For this to be tracked properly as a tool execution and to work properly in the tool form, I've implemented a new tool framework and tool form parameter type called "rules".

This can thought of as a more GUI friendly alternative to my proposed collection operations that consumed JavaScript expressions.

This includes API tests for both tool and workflow execution of the new tool as well as Selenium tests for tool form execution, workflow editor interactions, and workflow running.

jmchilton committed:
Fix selenium tests failing because APPLY_RULES appears twice in panel.

martenson commented:
conflicts because of https://github.com/galaxyproject/galaxy/pull/5969 merge - sorry 😭

jmchilton committed:
Merge other recent rule PRs into apply_rules branch.

martenson commented on client/galaxy/scripts/mvc/rules/rule-definitions.js:
```diff
@@ -713,7 +713,7 @@ const MAPPING_TARGETS = {
             "If this is set, all rows with the same collection name will be joined into a collection and it is possible to create multiple collections at once."
         ),
-        modes: ["raw", "ftp"], // TODO: allow this in datasets mode & tool builder modes
+        modes: ["raw", "ftp", "datasets", "library_datasets"],
```
you deleted this whole line in https://github.com/galaxyproject/galaxy/pull/5969, is this the proper merge?

jmchilton replied:
Yes... thanks for looking so close though! Think of modes like a filter - in #5969 I made it so all the modes of the rule builder that produce collections could production multiple collections so I didn't need that any more. This PR adds the ability to apply to existing collections using a tool that only has one output - so I needed to re-list all the valid modes and exclude that one new mode ("collection_contents").

<img width="825" alt="screen shot 2018-04-24 at 4 14 41 pm" src="https://user-images.githubusercontent.com/216771/39211891-0358ffb0-47db-11e8-9617-02fd2a57b194.png">
